Add VALGRIND.md to document valgrind complains (#29966)
openssl/openssl: NOTES-VALGRIND.md

    ==280439== Using Valgrind-3.24.0 and LibVEX; rerun with -h for copyright info
    ==280439== Command: ./pkeyread -f pem -k dh 8
    ==280439==
    Average time per pem(dh) call: 506329.113924us
    ==280439==
    ==280439== HEAP SUMMARY:
    ==280439==     in use at exit: 239,521 bytes in 4,137 blocks
    ==280439==   total heap usage: 21,841 allocs, 17,704 frees, 4,089,104 bytes allocated
    ==280439==
    ==280439== LEAK SUMMARY:
    ==280439==    definitely lost: 0 bytes in 0 blocks
    ==280439==    indirectly lost: 0 bytes in 0 blocks
    ==280439==      possibly lost: 0 bytes in 0 blocks
    ==280439==    still reachable: 239,521 bytes in 4,137 blocks
    ==280439==         suppressed: 0 bytes in 0 blocks
    ==280439== Rerun with --leak-check=full to see details of leaked memory
    ==280439==
    ==280439== For lists of detected and suppressed errors, rerun with: -s
    ==280439== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)

The valgrind output above reports there are 239,521 of reachable memory
when process exits. OpenSSL project sees that memory as not a memory leak.
The process is going to exit anyway and OS will reclaim that memory way

> faster than `OPENSSL_cleanup()` making calls to libc `free()`. Also

Reviewer: Hm, I'd amend this suggested change by wording the last part as "as the OS will reclaim that memory on process exit, rendering calls to libc `free()` within `OPENSSL_cleanup()` useless."

call `to OPENSSL_cleanup()` is discouraged when libcrypto is being linked
with process to satisfy more than one dependency paths. If it is the
case then call to `OPENSS_cleanup()` may lead to spurious application
crashes during exit.

If memory leaks caused by _still reachable memory_ are still issue
in your build/ci environment, then preferred way is to suppress

> those reports using suppression file [1] instead of changing exiting

Reviewer: `s/suppression/the suppression/`

code by adding call to `OPENSSL_cleanup()`.  The suppression file for

> OpenSSL is shipped with OpenSSL sources. It is found at

Reviewer 2: ```suggestion
OpenSSL is shipped within the OpenSSL sources and can be found at
```

> `$OPENSSL_SRCS/util/valgrind.suppressions`. To use it just add

Reviewer 2: ```suggestion
`$OPENSSL_SRCS/util/valgrind.suppressions` where `OPENSSL_SRCS`
is an environment variable containing path to the OpenSSL source
tree. To use it just add
```

> `--suppression` option to the valgrind command:

Reviewer 2: ```suggestion
`--suppressions` option to the valgrind command:
```

> `valgrind --suppressions=`OPENSSL_SRCS/util/valgrind.suppression ...`

Reviewer 2: ```suggestion
`valgrind --suppressions="$OPENSSL_SRCS/util/valgrind.suppression" ...`
```
